src/Connection/ConnectionLimitingPool.php

```diff
@@ -170,6 +176,14 @@ private function getStreamFor(string $uri, Request $request, CancellationToken $
                 $stream = yield $this->getStreamFromConnection($connection, $request);
 
                 if ($stream === null) {
+                    if (!$this->isAdditionalConnectionAllowed($uri)
+                        && $this->activeRequestCounts[\spl_object_id($connection)] === 0
```

Comment:
Maybe add an assertion that `$this->activeRequestCounts[\spl_object_id($connection)] >= 0`? I think it also needs a `?? 0`, because the connection might close during `getStreamFromConnection` above.